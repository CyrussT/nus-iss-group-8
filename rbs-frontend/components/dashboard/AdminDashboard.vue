<script setup lang="ts">
import { ref, onMounted } from "vue";
import axios from "axios";

const todayBookingsCount = ref(0);
const pendingApprovalsCount = ref(0);
const facilitiesUnderMaintenanceCount = ref(0);
const emergencyMessage = ref<string | null>(null);

const fetchAdminStats = async () => {
  try {
    const res = await axios.get('http://localhost:8080/api/bookings/admin/dashboard-stats');
    const data = res.data;
    todayBookingsCount.value = data.todayBookings;
    pendingApprovalsCount.value = data.pendingApprovals;
    facilitiesUnderMaintenanceCount.value = data.facilitiesUnderMaintenance;
    emergencyMessage.value = data.emergencyMessage || null;
  } catch (error) {
    console.error('Error fetching admin stats', error);
  }
};

onMounted(fetchAdminStats);
</script>

<template>
  <div class="p-8 flex flex-col gap-6">
    <h1 class="text-2xl font-bold">Admin Dashboard</h1>
    <div class="grid grid-cols-1 md:grid-cols-3 gap-6">
      <div class="bg-white p-6 rounded-lg shadow-md text-center">
        <p class="text-gray-600">Today's Bookings</p>
        <p class="text-4xl font-bold">{{ todayBookingsCount }}</p>
      </div>
      <div class="bg-white p-6 rounded-lg shadow-md text-center">
        <p class="text-gray-600">Pending Approvals</p>
        <p class="text-4xl font-bold">{{ pendingApprovalsCount }}</p>
      </div>
      <div class="bg-white p-6 rounded-lg shadow-md text-center">
        <p class="text-gray-600">Facilities Under Maintenance</p>
        <p class="text-4xl font-bold">{{ facilitiesUnderMaintenanceCount }}</p>
      </div>
    </div>
    <div v-if="emergencyMessage" class="bg-red-100 border border-red-300 p-4 rounded-lg mt-6">
      <h2 class="font-bold text-lg mb-2">ðŸš¨ Emergency Message</h2>
      <p>{{ emergencyMessage }}</p>
    </div>
  </div>
</template>
