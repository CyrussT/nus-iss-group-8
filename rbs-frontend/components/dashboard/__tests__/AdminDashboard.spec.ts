import { describe, it, expect, vi, beforeEach } from 'vitest';
import { mount } from '@vue/test-utils';
import AdminDashboard from '../AdminDashboard.vue';
import axios from 'axios';

// Mock axios
vi.mock('axios');

// Mock composables
vi.mock('@/composables/useApi', () => ({
  useApi: () => ({
    apiUrl: 'http://localhost:8080'
  })
}));

describe('AdminDashboard', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    
    // Set up axios mock response
    (axios.get as any).mockResolvedValue({
      data: {
        todayBookings: 5,
        pendingApprovals: 2,
        facilitiesUnderMaintenance: 1,
        emergencyMessage: null
      }
    });
  });
  
  it('renders correctly with stats', async () => {
    const wrapper = mount(AdminDashboard, {
      global: {
        stubs: {
          UButton: true,
          UCard: true,
          UModal: true,
          UTextarea: true
        }
      }
    });
    
    // Wait for the component to update after API call
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    
    expect(wrapper.html()).toMatchSnapshot();
    
    // Verify stats are displayed
    expect(wrapper.find('p.text-4xl.font-bold.text-center').text()).toBe('5');
    expect(wrapper.findAll('p.text-4xl.font-bold.text-center')[1].text()).toBe('2');
    expect(wrapper.findAll('p.text-4xl.font-bold.text-center')[2].text()).toBe('1');
  });
  
  it('renders emergency message when present', async () => {
    // Mock response with emergency message
    (axios.get as any).mockResolvedValue({
      data: {
        todayBookings: 10,
        pendingApprovals: 3,
        facilitiesUnderMaintenance: 2,
        emergencyMessage: "Emergency: System maintenance at 5PM"
      }
    });
    
    const wrapper = mount(AdminDashboard, {
      global: {
        stubs: {
          UButton: true,
          UCard: true,
          UModal: true,
          UTextarea: true
        }
      }
    });
    
    // Wait for the component to update after API call
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    
    expect(wrapper.html()).toMatchSnapshot();
    
    // Verify emergency message is displayed
    const emergencyCard = wrapper.find('p.text-red-700');
    expect(emergencyCard.exists()).toBe(true);
    expect(emergencyCard.text()).toBe("Emergency: System maintenance at 5PM");
  });
  
  it('opens emergency modal when button is clicked', async () => {
    const wrapper = mount(AdminDashboard, {
      global: {
        stubs: {
          UButton: false,
          UCard: true,
          UModal: {
            template: '<div v-if="modelValue"><slot /></div>',
            props: ['modelValue']
          },
          UTextarea: true
        }
      }
    });
    
    // Emergency modal should be closed initially
    expect(wrapper.vm.isEmergencyModalOpen).toBe(false);
    
    // Click the emergency button
    await wrapper.find('button.px-4.py-2').trigger('click');
    
    // Modal should be open
    expect(wrapper.vm.isEmergencyModalOpen).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
