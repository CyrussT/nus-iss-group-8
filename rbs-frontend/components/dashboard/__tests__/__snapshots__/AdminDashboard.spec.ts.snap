// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdminDashboard > opens emergency modal when button is clicked 1`] = `
"<div class="p-8 space-y-6">
  <!-- Header and Emergency Button -->
  <div class="flex justify-between items-center">
    <h1 class="text-2xl font-bold">Admin Dashboard</h1><button type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm text-white dark:text-gray-900 bg-red-500 hover:bg-red-600 disabled:bg-red-500 aria-disabled:bg-red-500 dark:bg-red-400 dark:hover:bg-red-500 dark:disabled:bg-red-400 dark:aria-disabled:bg-red-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-red-500 dark:focus-visible:outline-red-400 inline-flex items-center px-4 py-2 gap-2">
      <!--v-if--><span class="">Push Emergency Notice</span>
      <!--v-if-->
    </button>
  </div><!-- Stat Cards -->
  <div class="grid grid-cols-1 md:grid-cols-3 gap-6">
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Today's Bookings</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">5</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Pending Approvals</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">2</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Facilities Under Maintenance</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">1</p>
      </div>
      <!--v-if-->
    </div>
  </div><!-- Emergency Message Display -->
  <!--v-if-->
  <!-- Emergency Modal -->
  <!--teleport start-->
  <!--teleport end-->
  <div hidden="" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0px; display: none;"></div>
</div>"
`;

exports[`AdminDashboard > renders correctly with stats 1`] = `
"<div class="p-8 space-y-6">
  <!-- Header and Emergency Button -->
  <div class="flex justify-between items-center">
    <h1 class="text-2xl font-bold">Admin Dashboard</h1><button type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm text-white dark:text-gray-900 bg-red-500 hover:bg-red-600 disabled:bg-red-500 aria-disabled:bg-red-500 dark:bg-red-400 dark:hover:bg-red-500 dark:disabled:bg-red-400 dark:aria-disabled:bg-red-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-red-500 dark:focus-visible:outline-red-400 inline-flex items-center px-4 py-2 gap-2">
      <!--v-if--><span class="">Push Emergency Notice</span>
      <!--v-if-->
    </button>
  </div><!-- Stat Cards -->
  <div class="grid grid-cols-1 md:grid-cols-3 gap-6">
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Today's Bookings</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">5</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Pending Approvals</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">2</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Facilities Under Maintenance</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">1</p>
      </div>
      <!--v-if-->
    </div>
  </div><!-- Emergency Message Display -->
  <!--v-if-->
  <!-- Emergency Modal -->
  <!---->
</div>"
`;

exports[`AdminDashboard > renders emergency message when present 1`] = `
"<div class="p-8 space-y-6">
  <!-- Header and Emergency Button -->
  <div class="flex justify-between items-center">
    <h1 class="text-2xl font-bold">Admin Dashboard</h1><button type="button" class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:cursor-not-allowed aria-disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm shadow-sm text-white dark:text-gray-900 bg-red-500 hover:bg-red-600 disabled:bg-red-500 aria-disabled:bg-red-500 dark:bg-red-400 dark:hover:bg-red-500 dark:disabled:bg-red-400 dark:aria-disabled:bg-red-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-red-500 dark:focus-visible:outline-red-400 inline-flex items-center px-4 py-2 gap-2">
      <!--v-if--><span class="">Push Emergency Notice</span>
      <!--v-if-->
    </button>
  </div><!-- Stat Cards -->
  <div class="grid grid-cols-1 md:grid-cols-3 gap-6">
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Today's Bookings</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">10</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Pending Approvals</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">3</p>
      </div>
      <!--v-if-->
    </div>
    <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow bg-white dark:bg-gray-900">
      <div class="px-4 py-5 sm:px-6">
        <p class="text-gray-500">Facilities Under Maintenance</p>
      </div>
      <div class="px-4 py-5 sm:p-6">
        <p class="text-4xl font-bold text-center">2</p>
      </div>
      <!--v-if-->
    </div>
  </div><!-- Emergency Message Display -->
  <div class="rounded-lg divide-y divide-gray-200 dark:divide-gray-800 ring-1 ring-gray-200 dark:ring-gray-800 shadow dark:bg-gray-900 border border-red-300 bg-red-50">
    <div class="px-4 py-5 sm:px-6">
      <h2 class="text-lg font-bold text-red-700">ðŸš¨ Emergency Message</h2>
    </div>
    <div class="px-4 py-5 sm:p-6">
      <p class="text-red-700">Emergency: System maintenance at 5PM</p>
    </div>
    <!--v-if-->
  </div><!-- Emergency Modal -->
  <!---->
</div>"
`;
